### PR TITLE
cephfs,mon: fix bugs related to updating MDS caps

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
https://tracker.ceph.com/issues/64127
https://tracker.ceph.com/issues/64182
https://tracker.ceph.com/issues/64417

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: qa/tasks/cephfs/test_admin.py [hunk 1]
```diff
@@ -1424,15 +1424,49 @@ def test_multiple_path_rw(self):
...
         for i, c in enumerate(fs_auth_caps):
             self.assertIn(i, (0, 1))
             PATH = c[0]
             PERM = c[1]
             self._remount(keyring, PATH)
             # actual tests...
-            self.captesters[i].run_cap_tests(self.fs, self.client_id, PERM,
-                                             PATH)
+            self.captesters[i].run_cap_tests(self.fs, self.client_id, PERM)
+            #self.captesters[i].run_cap_tests(self.fs, self.client_id, PERM,
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:1479]: why is this commented out?

File: qa/tasks/cephfs/test_admin.py [hunk 2]
```diff
@@ -1424,15 +1424,49 @@ def test_multiple_path_rw(self):
...
+
+        The first command will create the keyring by adding the MDS cap for
+        root path & MON and OSD caps with name of the FS name (say "cephfs1").
+        But second and third command don't need to add MON and OSD caps since
+        cap for "cephfs2" has been granted already. Thus, second and third
+        command only need to update the MDS cap, thus testing the bug
+        under consideration here.
+        '''
+        PERM = 'rw'
+        FS_AUTH_CAPS = (('/', PERM), ('dir1', PERM), ('/dir2', PERM))
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:1448]: Should this be `/dir1` or maybe just "dir2", but not both.

File: src/mds/MDSAuthCaps.cc [hunk 3]
```diff
@@ -382,25 +382,29 @@ bool MDSAuthCaps::parse(string_view str, ostream *err)
   }
 }
 
-bool MDSAuthCaps::merge(MDSAuthCaps newcap)
+/* Check if the "cap grant" is already present in this cap object. If it is,
+ * return false. If not, add it and return true.
+ *
+ * ng = new grant, new mds cap grant, shorter name makes this code more
+ * succinct.
```
Comment (vshankar) [inline src/mds/MDSAuthCaps.cc:389]: really? Code is succinct when it's a) cleanly/clearly written and b) adequately commented. I disagree!!

File: src/mds/MDSAuthCaps.cc [hunk 4]
```diff
@@ -409,6 +413,39 @@ bool MDSAuthCaps::merge(MDSAuthCaps newcap)
...
+ * AuthMonitor runs the update procedure.
+ */
+bool MDSAuthCaps::merge(MDSAuthCaps newcaps)
+{
+  vector<bool> were_caps_merged;
+
+  for (auto& ng : newcaps.grants) {
+      bool is_merged = merge_one_cap_grant(ng);
+      were_caps_merged.push_back(is_merged);
+  }
```
Comment (vshankar) [inline src/mds/MDSAuthCaps.cc:435]: A slightly better version:

```
bool merged = false;

for (auto& ng : newcaps.grants) {
        merged |= merge_one_cap_grant(ng);
}

return merged;

**Review by vshankar — CHANGES_REQUESTED** [review]

**Comment by vshankar** [comment]
@rishabh-d-dave I don't think this change needs backporting but please double check: https://tracker.ceph.com/issues/64127#note-5

File: src/mds/MDSAuthCaps.cc (hunk 3 above)
Comment (rishabh-d-dave) [reply src/mds/MDSAuthCaps.cc:389]: Ok. Removed.

File: src/mds/MDSAuthCaps.cc (hunk 4 above)
Comment (rishabh-d-dave) [reply src/mds/MDSAuthCaps.cc:435]: Much better and simpler, thanks!

**Comment by rishabh-d-dave** [comment]
https://github.com/ceph/ceph/compare/918e2f49e3e50202429265fa8ed02645cf2d6616..dbcf868a740fe61d340e8831d3c8149c0e915244

https://github.com/ceph/ceph/compare/dbcf868a740fe61d340e8831d3c8149c0e915244..b0b38b1a2ebcbbd0f91bccf8c54db38cdbaae783

https://github.com/ceph/ceph/compare/b0b38b1a2ebcbbd0f91bccf8c54db38cdbaae783..77a86c1719c0ff58bcdac68aba14f720517b7a45

File: qa/tasks/cephfs/test_admin.py (hunk 2 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:1448]: This was too ensure that adding multiple caps for first FS works, which was seen during investigating this bug.

File: qa/tasks/cephfs/test_admin.py (hunk 1 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:1490]: This part wasn't tested. Needed to take care of it eventually.

Comment (vshankar) [reply qa/tasks/cephfs/test_admin.py:1490]: Sure, please remove it if its unneeded.

File: qa/tasks/cephfs/test_admin.py (hunk 2 above)
Comment (vshankar) [reply qa/tasks/cephfs/test_admin.py:1448]: Sure, but I'm pointing to `dir1` and `/dir2` (one with `/` and the other without it).

File: src/mon/AuthMonitor.cc [hunk 5]
```diff
@@ -1888,15 +1891,32 @@ AuthMonitor::caps_update AuthMonitor::_gen_wanted_caps(EntityAuth& e_auth,
...
+    is_caps_update_reqd = CAPS_PARSING_ERR;
+  // even if any one of MON, OSD or MDS caps needs an update, the update
+  // procedure needs to be executed.
+  } else if (is_caps_update_reqd_mon == CAPS_UPDATE_REQD ||
+      is_caps_update_reqd_osd == CAPS_UPDATE_REQD ||
+      is_caps_update_reqd_mds == CAPS_UPDATE_REQD) {
+    is_caps_update_reqd = CAPS_UPDATE_REQD;
+  // only if all caps don't require update, don't run the update procedure.
+  } else {
+    is_caps_update_reqd = CAPS_UPDATE_NOT_REQD;
```
Comment (vshankar) [inline src/mon/AuthMonitor.cc:1919]: Is the `else` part needed - the default value for `is_caps_update_reqd` is `CAPS_UPDATE_NOT_REQD`.

File: qa/tasks/cephfs/test_admin.py (hunk 1 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:1490]: Done locally. I'll update the PR shortly.

File: qa/tasks/cephfs/test_admin.py (hunk 2 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:1448]: Ah, I suppose you mean either to have `/dir1` and `/dir2` or to have `dir1` and `dir2`...

Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py:1448]: Anyways, this part is not vital to the test case, so I've removed it/

File: src/mon/AuthMonitor.cc (hunk 5 above)
Comment (rishabh-d-dave) [reply src/mon/AuthMonitor.cc:1919]: True, but this is just to be safe. In future, if there is a change between first line and last line due to which value of `is_caps_update_reqd` is disturbed, this line would save us from a bug.

Comment (vshankar) [reply src/mon/AuthMonitor.cc:1919]: That code would be vetted during review, so , I don't think its required.

File: qa/tasks/cephfs/test_admin.py (hunk 1 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py]: Done. Marking this thread resolved.

File: qa/tasks/cephfs/test_admin.py (hunk 2 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_admin.py]: Done. Marking this thread resolved.

**Comment by rishabh-d-dave** [comment]
.

**Comment by rishabh-d-dave** [comment]
jenkins test api

File: src/mon/AuthMonitor.cc (hunk 5 above)
Comment (vshankar) [reply src/mon/AuthMonitor.cc:1919]: Why isn't this taken care of @rishabh-d-dave esp when you requested a review of the changes?

Comment (rishabh-d-dave) [reply src/mon/AuthMonitor.cc]: It's easy to miss such cases during review, so I thought it is good to have. But, anyways, I've removed it now.

Comment (vshankar) [reply src/mon/AuthMonitor.cc]: > It's easy to miss such cases during review, so I thought it is good to have. But, anyways, I've removed it now.

Sure. But I'd expect to be explicitly mentioned in the comment reply that its not been taken care of, else to me it looks like the change request has been ignored.

**Comment by vshankar** [comment]
Lgtm

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20240226.124304

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2024-02-27_04:05:06-fs-wip-vshankar-testing-20240226.124304-testing-default-smithi/

**Comment by vshankar** [comment]
Reruning tests with https://github.com/ceph/ceph/pull/55909 and https://github.com/ceph/ceph/pull/55908 included. See - https://pulpito.ceph.com/vshankar-2024-03-04_08:26:39-fs-wip-vshankar-testing-20240304.042522-testing-default-smithi/

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#5th-March-2024

**Comment by vshankar** [comment]
@rzarzynski This change touches src/mon/AuthMonitor.cc. Do you want to review the changes? (passes fs suite).

**Comment by rishabh-d-dave** [comment]
> @rzarzynski This change touches src/mon/AuthMonitor.cc. Do you want to review the changes? (passes fs suite).

@ljflores

**Comment by rishabh-d-dave** [comment]
Confirmed with @vshankar that it's okay to merge this PR.

**Comment by rishabh-d-dave** [comment]
> https://tracker.ceph.com/projects/cephfs/wiki/Main#5th-March-2024

QA was successful. Relinking this comment here.

**Comment by rishabh-d-dave** [comment]
> @vshankar IMO this PR should be backported. Any thoughts? Copying link to the related ticket below -
> 
> https://tracker.ceph.com/issues/64127 https://tracker.ceph.com/issues/64182 https://tracker.ceph.com/issues/64417

@vshankar ping

**Comment by vshankar** [comment]
> > @vshankar IMO this PR should be backported. Any thoughts? Copying link to the related ticket below -
> > 
> > https://tracker.ceph.com/issues/64127 https://tracker.ceph.com/issues/64182 https://tracker.ceph.com/issues/64417
> 
> @vshankar ping

If the original changes have been backported, then yes, this change needs to be backported.

**Pull request merged**